### ISSUE TITLE
Limit the completion to just those variables of the appropriate enum types.

### ISSUE THREAD

**Issue by hongyi-zhao** [python-lsp/python-lsp-server#101]
On Ubuntu 20.04.3 LTS, I'm using self-compiled git master version of Emacs with `python-lsp-server`. See the following code completion results given by `python-lsp-server`:

![image](https://user-images.githubusercontent.com/11155854/139530967-d39251d5-228d-4578-bde4-ede223db8e1f.png)

As you can see, company gives so many undesired/incorrect completion candidates in the above screenshot. According the docstring [here](https://github.com/python/cpython/blob/62bf263a775f4444d8b5d5841cc09be3bd53e933/Lib/re.py#L103-L118), only the following flags are allowed:

```
    A  ASCII       For string patterns, make \w, \W, \b, \B, \d, \D
                   match the corresponding ASCII character categories
                   (rather than the whole Unicode categories, which is the
                   default).
                   For bytes patterns, this flag is the only available
                   behaviour and needn't be specified.
    I  IGNORECASE  Perform case-insensitive matching.
    L  LOCALE      Make \w, \W, \b, \B, dependent on the current locale.
    M  MULTILINE   "^" matches the beginning of lines (after a newline)
                   as well as the string.
                   "$" matches the end of lines (before a newline) as well
                   as the end of the string.
    S  DOTALL      "." matches any character at all, including the newline.
    X  VERBOSE     Ignore whitespace and comments for nicer looking RE's.
    U  UNICODE     For compatibility only. Ignored for string patterns (it
                   is the default), and forbidden for bytes patterns.
```
So, I want to know the suggested company-backends setting to obtain desirable completion candidates when using `python-lsp-server`. Also see the discussion [here](https://github.com/company-mode/company-mode/discussions/1256) for relevant discussion.

Regards,
HZ

**Comment by lieryan** [comment]
This is nothing to do with completion settings, but all to do with having the right type hints that tells the completer what values `re.compile(..., flags)` takes.

In this case `flags` are enums and in theory the `re` module could add a type hint indicating which enum class is valid values for the `flags` parameter and the completer could use that info to limit the completion to just those variables of the appropriate enum types.

python-lsp-server can use either `jedi` or `rope` library for completion; but by default it uses `jedi`. I don't think hinting enums in this way is currently supported in `rope` (I'm a `rope` maintainer). You may want to look into `jedi` documentation and ask them if they support this kind of completion.

**Comment by PeterJCLaw** [comment]
This is theoretically possible (`jedi` provides the raw type information from the typeshed), and I can see why you're asking this, however the resulting behaviour is highly unlikely to actually be what you want. Consider the following use-case, which what you're asking for would make harder to spell:

``` python
SETTINGS = re.IGNORECASE | re.MULTILINE
...
re.match('foo', 'bar', SETTINGS)
     # completion here ^
```

If in the completion of the last parameter here only `re.XYZ` completions were offered, then `SETTINGS` would be missing (and surprisingly so). While it's possible to augment the result to detect and allow constants, since an item of almost any type can be reached through member access of any object, the valid (and thus potentially useful) chain of symbol completions at any point is almost always all valid symbols in scope.

As a result the augmentations needed to allow things which _might_ be valid would involve checking all possible member access, which is likely to be highly non-performant and yet not likely to reduce the candidate pool of suggestions meaningfully.

The alternative might be that we only apply this limiting for members of `re`. However the question is then whether `from re import IGNORECASE` style symbol access should be filtered? An approach of this sort (limited base on the symbol source) would also require that the implementation be filled with special cases for each callsite which needs this behaviour, which would become a maintenance burden pretty quickly and (as noted) offers only marginal value.

There definitely are places where a more limited pool of suggestions is practical, these are relatively few and far between.

<details>
<summary>Footnote: a case where a limited pool is practical</summary>

Consider the literal keys of a `TypedDict`, which must be literals and must come from the valid keys of the `TypedDict` and thus sidestep the above issues, though note that Jedi has thus far not implemented this on the grounds of not being worth the (non-trivial) effort of doing so.

</details>

**Comment by krassowski** [comment]
Thanks for chiming in! Very good point on that reliable filtering is likely out of scope for both `jedi` and `pylsp` for reasons outlined above; however, what might be more useful would be trying to improve sorting to place the _known_ good matches at the top. It is a non-trivial problem, and frankly easier to solve with frequency analysis of existing code (currently sold as _ML_), but I think we should be open to explore our options here.

**Comment by PeterJCLaw** [comment]
Indeed, using this sort of information to improve sorting strikes me as a much more fruitful direction :+1:. (There remain challenges around avoiding lots of special casing though.)

**Comment by hongyi-zhao** [comment]
Another relevant [problem](https://github.com/tumashu/company-posframe/issues/55) is the incorrect docstring retured by `company-posframe`.
